bblfsh/pyuast.cc

```diff
@@ -1,272 +1,760 @@
+// Returns a new reference.
+static PyObject *PythonContextExt_root(PythonContextExt *self, PyObject *Py_UNUSED(ignored)) {
+    return self->p->RootNode();
+}
 
-  static uint32_t StartLine(const void *node) {
-    return PositionValue(node, "start_position", "line");
-  }
+// PythonContextExt_load returns a root node converted to Python object.
+// Returns a new reference.
+static PyObject *PythonContextExt_load(PythonContextExt *self, PyObject *Py_UNUSED(ignored)) {
+    PyObject* root = PythonContextExt_root(self, nullptr);
+    return PyNodeExt_load((PyNodeExt*)root, nullptr);
+}
 
-  static bool HasStartCol(const void *node) {
-    return HasAttribute(node, "start_position");
-  }
+// PythonContextExt_filter filters UAST.
+// Returns a new reference.
+static PyObject *PythonContextExt_filter(PythonContextExt *self, PyObject *args, PyObject *kwargs) {
+    char* kwds[] = {(char*)"query", (char*)"node", NULL};
+    const char *query = nullptr;
+    PyObject *node = nullptr;
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "s|O", kwds, &query, &node))
+      return nullptr;
+
+    PyObject* it = nullptr;
+    try {
+        it = self->p->Filter(node, query);
+    } catch (const std::exception& e) {
+        PyErr_SetString(PyExc_RuntimeError, e.what());
+    }
+    return it;
+}
 
-  static uint32_t StartCol(const void *node) {
-    return PositionValue(node, "start_position", "col");
-  }
+// PythonContextExt_filter serializes UAST.
+// Returns a new reference.
+static PyObject *PythonContextExt_encode(PythonContextExt *self, PyObject *args) {
+    PyObject *node = nullptr;
+    UastFormat format = UAST_BINARY; // TODO: make it a kwarg and enum
+    if (!PyArg_ParseTuple(args, "Oi", &node, &format))
+      return nullptr;
+    return self->p->Encode(node, format);
+}
 
-  static bool HasEndOffset(const void *node) {
-    return HasAttribute(node, "end_position");
-  }
+static PyMethodDef PythonContextExt_methods[] = {
+    {"root", (PyCFunction) PythonContextExt_root, METH_NOARGS,
+     "Return the root node attached to this query context"
+    },
+    {"load", (PyCFunction) PythonContextExt_load, METH_NOARGS,
+     "Load external node to Python"
+    },
+    {"filter", (PyCFunction) PythonContextExt_filter, METH_VARARGS | METH_KEYWORDS,
+     "Filter a provided UAST with XPath"
+    },
+    {"encode", (PyCFunction) PythonContextExt_encode, METH_VARARGS,
+     "Encodes a UAST into a buffer"
+    },
+    {nullptr}  // Sentinel
+};
 
-  static uint32_t EndOffset(const void *node) {
-    return PositionValue(node, "end_position", "offset");
-  }
+extern "C"
+{
+    static PyTypeObject PythonContextExtType = {
+      PyVarObject_HEAD_INIT(nullptr, 0)
+      "pyuast.ContextExt",               // tp_name
+      sizeof(PythonContextExt),                 // tp_basicsize
+      0,                              // tp_itemsize
+      PythonContextExt_dealloc,                 // tp_dealloc
```

Comment:
Here, above and in some other `PyTypeObject`s formatting seems to be a bit off.

Reply:
Fixed in next commit.